crates/common/validator/src/sync_committee.rs

```diff
@@ -0,0 +1,85 @@
+use std::collections::HashSet;
+
+use anyhow::{anyhow, ensure};
+use ream_consensus::{
+    constants::{EPOCHS_PER_SYNC_COMMITTEE_PERIOD, SYNC_COMMITTEE_SIZE},
+    electra::beacon_state::BeaconState,
+    misc::compute_epoch_at_slot,
+};
+
+use crate::constants::SYNC_COMMITTEE_SUBNET_COUNT;
+
+pub fn compute_sync_committee_period(epoch: u64) -> anyhow::Result<u64> {
+    Ok(epoch / EPOCHS_PER_SYNC_COMMITTEE_PERIOD)
+}
+
+pub fn is_assigned_to_sync_committee(
+    state: &BeaconState,
+    epoch: u64,
+    validator_index: u64,
+) -> anyhow::Result<bool> {
+    let sync_committee_period = compute_sync_committee_period(epoch)?;
+    let current_epoch = state.get_current_epoch();
+    let current_sync_committee_period = compute_sync_committee_period(current_epoch)?;
+    let next_sync_committee_period = current_sync_committee_period + 1;
+    ensure!(
+        sync_committee_period == current_sync_committee_period
+            || sync_committee_period == next_sync_committee_period,
+        "Validator is not assigned to sync committee for period {sync_committee_period} (current: {current_sync_committee_period}, next: {next_sync_committee_period})"
+    );
+
+    let validator_pubkey = state
+        .validators
+        .get(validator_index as usize)
+        .ok_or_else(|| anyhow!("Validator index out of bounds"))?
+        .pubkey
+        .clone();
+
+    if sync_committee_period == current_sync_committee_period {
+        Ok(state
+            .current_sync_committee
+            .pubkeys
+            .contains(&validator_pubkey))
+    } else {
+        Ok(state
+            .next_sync_committee
+            .pubkeys
+            .contains(&validator_pubkey))
+    }
+}
+
+pub fn compute_subnets_for_sync_committee(
+    state: &BeaconState,
+    validator_index: u64,
+) -> anyhow::Result<HashSet<u8>> {
+    let next_slot_epoch = compute_epoch_at_slot(state.slot + 1);
+    let sync_committee = if compute_sync_committee_period(state.get_current_epoch())?
+        == compute_sync_committee_period(next_slot_epoch)?
+    {
+        &state.current_sync_committee
+    } else {
+        &state.next_sync_committee
+    };
+
+    let target_pubkey = state
+        .validators
+        .get(validator_index as usize)
+        .ok_or_else(|| anyhow!("Validator index out of bounds"))?
+        .pubkey
+        .clone();
+
+    let sync_committee_indices: Vec<usize> = sync_committee
+        .pubkeys
+        .iter()
+        .enumerate()
+        .filter(|(_, pubkey)| **pubkey == target_pubkey)
+        .map(|(index, _)| index)
+        .collect();
```

Comment:
```suggestion
    let target_pubkey = state
        .validators
        .get(validator_index as usize)
        .map(|validator| &validator.pubkey)
        .ok_or_else(|| anyhow!("Validator index out of bounds"))?;

    let sync_committee_indices: Vec<usize> = sync_committee
        .pubkeys
        .iter()
        .enumerate()
        .filter_map(|(index, pubkey)| {
            if pubkey == target_pubkey {
                Some(index)
            } else {
                None
            }
        })
        .collect();
```

I think you can use `filter_map`

Reply:
I think the filter map is less clean in this case